scope_test.go

```diff
@@ -0,0 +1,180 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dig
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestScopedOperations(t *testing.T) {
+	t.Parallel()
+
+	t.Run("getStores/ScopesFromRoot returns scopes from root in order of distance from root", func(t *testing.T) {
+		c := New()
+		s1 := c.Scope("child1")
+		s2 := s1.Scope("child2")
+		s3 := s2.Scope("child2")
+
+		assert.Equal(t, []containerStore{c.scope, s1, s2, s3}, s3.getStoresFromRoot())
+		assert.Equal(t, []*Scope{c.scope, s1, s2, s3}, s3.getScopesFromRoot())
+	})
+
+	t.Run("private provides", func(t *testing.T) {
+		c := New()
+		s := c.Scope("child")
+		type A struct{}
+
+		f := func(a *A) {
+			assert.NotEqual(t, nil, a)
+		}
+
+		require.NoError(t, s.Provide(func() *A { return &A{} }))
+		assert.NoError(t, s.Invoke(f))
+		assert.Error(t, c.Invoke(f))
+	})
+
+	t.Run("private provides inherits", func(t *testing.T) {
+		type A struct{}
+		type B struct{}
+
+		useA := func(a *A) {
+			assert.NotEqual(t, nil, a)
+		}
+		useB := func(b *B) {
+			assert.NotEqual(t, nil, b)
+		}
+
+		c := New()
+		c.Provide(func() *A { return &A{} })
+
+		child := c.Scope("child")
+		child.Provide(func() *B { return &B{} })
+		assert.NoError(t, child.Invoke(useA))
+		assert.NoError(t, child.Invoke(useB))
+
+		grandchild := child.Scope("grandchild")
+
+		assert.NoError(t, grandchild.Invoke(useA))
+		assert.NoError(t, grandchild.Invoke(useB))
+		assert.Error(t, c.Invoke(useB))
+	})
+
+	t.Run("provides to top-level Container propogates to all scopes", func(t *testing.T) {
+		type A struct{}
+
+		// Scope tree:
+		//     root  <-- Provide(func() *A)
+		//    /    \
+		//   c1	    c2
+		//   |     /  \
+		//   gc1  gc2  gc3
+		var allScopes []*Scope
+		root := New()
+
+		allScopes = append(allScopes, root.Scope("child 1"), root.Scope("child 2"))
+		allScopes = append(allScopes, allScopes[0].Scope("grandchild 1"), allScopes[1].Scope("grandchild 2"), allScopes[1].Scope("grandchild 3"))
+
+		root.Provide(func() *A {
+			return &A{}
+		})
+
+		// top-level provide should be available in all the scopes.
+		for _, scope := range allScopes {
+			assert.NoError(t, scope.Invoke(func(a *A) {}))
+		}
+	})
+
+	t.Run("private provides to child should be available to grandchildren, but not root", func(t *testing.T) {
```

Comment:
hmm.. yeah, im not sure whether this test is a true positive or a negative and it should be grouped below. leaning towards putting this into `TestScopeFailures` but just bringing this up to your attention.

Reply:
hmm, you're right. this should probably go to failure case. I'll move it now. Thanks for catching it!